### PR TITLE
Bridge to handle port mapping

### PR DESCRIPTION
**Pull request by aboch**
- libnetwork cares for list of exposed ports, driver cares
  for list of port bindings:
  - list of exposed ports will be passed during Join as option
  - list of port mapping will be passed during Join as driver config

- in endpoint.go made parameters of unexported structs unexported

- taken care of mavenugo and mrjana's comments

Signed-off-by: Alessandro Boch <aboch@docker.com>

### CODE REVIEW

**Comment by tomdee** [comment]
How could a plugin be written to do these port mappings? Could the bridge be written as a plugin (replaceable batteries) and still do these port mappings?

**Comment by mavenugo** [comment]
@tomdee Yes. The driver configuration that is mentioned in the PR is a well-known config that any driver can choose to implement. (Some Drivers may not implement and even error out if that is not applicable).
This PR might need some minor tweaks to accommodate this.

**Comment by squaremo** [comment]
> The driver configuration that is mentioned in the PR is a well-known config that any driver can choose to implement.

What's the contract here: what should the user (or the application on behalf of the user) expect to be the case if they pass this port option?

**Comment by mrjana** [comment]
@squaremo This PR need to be updated as per the contract. But the contract is driver will get some config which are tagged with well-known labels (as defined in driverapi) which they should mandatorily process. If these configs are not applicable to the driver it should return an error saying so. 

This way options such as port mapping/ mac address etc. can be passed by the user on containers on any network implemented by any driver but if they try to add a container to a network implemented by a particular driver which does not accept that particular option then they will get an appropriate error notification and container not allowed to join that network.

File: drivers/bridge/bridge.go [hunk 1]
```diff
@@ -404,10 +421,11 @@ func (d *driver) CreateEndpoint(nid, eid types.UUID, epOptions map[string]interf
...
-	if epConfig != nil && epConfig.MacAddress != nil {
-		mac = epConfig.MacAddress
+	// Set the sbox's MAC. If specified, use the one configured by user, otherwise use a random one
+	mac := electMacAddress(epConfig)
+	err = netlink.LinkSetHardwareAddr(sbox, mac)
+	if err != nil {
+		return nil, err
 	}
 
 	err = netlink.LinkSetHardwareAddr(sbox, mac)
```
Comment (mavenugo) [inline drivers/bridge/bridge.go]: Why is a duplicate call to `LinkSetHardwareAddr` added ?

Comment (aboch) [inline drivers/bridge/bridge.go]: It's a bug. Thanks for catching this.

**Comment by mavenugo** [comment]
LGTM.

**Comment by mrjana** [comment]
LGTM

**Pull request merged**